ast/src/analyzed/mod.rs

```diff
@@ -715,96 +710,240 @@ impl<Expr> Default for SelectedExpressions<Expr> {
     }
+}
 
-    pub fn contains_next_ref(&self) -> bool {
-        self.left.contains_next_ref() || self.right.contains_next_ref()
+#[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize, JsonSchema)]
+pub struct ConnectIdentity<T> {
+    // The ID is globally unique among identitites.
+    pub id: u64,
+    pub source: SourceRef,
+    pub left: Vec<AlgebraicExpression<T>>,
+    pub right: Vec<AlgebraicExpression<T>>,
+}
+
+impl<T> Children<AlgebraicExpression<T>> for ConnectIdentity<T> {
+    fn children_mut(&mut self) -> Box<dyn Iterator<Item = &mut AlgebraicExpression<T>> + '_> {
+        Box::new(
+            self.left
+                .iter_mut()
+                .chain(self.right.iter_mut())
+                .flat_map(|e| e.children_mut()),
+        )
+    }
+    fn children(&self) -> Box<dyn Iterator<Item = &AlgebraicExpression<T>> + '_> {
+        Box::new(
+            self.left
+                .iter()
+                .chain(self.right.iter())
+                .flat_map(|e| e.children()),
+        )
     }
+}
 
-    /// Either returns (a, Some(b)) if this is a - b or (a, None)
-    /// if it is a polynomial identity of a different structure.
-    /// Panics if it is a different kind of constraint.
-    pub fn as_polynomial_identity(
-        &self,
-    ) -> (&AlgebraicExpression<T>, Option<&AlgebraicExpression<T>>) {
-        assert_eq!(self.kind, IdentityKind::Polynomial);
-        match self.expression_for_poly_id() {
-            AlgebraicExpression::BinaryOperation(AlgebraicBinaryOperation {
-                left: a,
-                op: AlgebraicBinaryOperator::Sub,
-                right: b,
-            }) => (a.as_ref(), Some(b.as_ref())),
-            a => (a, None),
+impl<T> ConnectIdentity<T> {
+    pub fn new(
+        id: u64,
+        source: SourceRef,
+        left: Vec<AlgebraicExpression<T>>,
+        right: Vec<AlgebraicExpression<T>>,
+    ) -> Self {
+        Self {
+            id,
+            source,
+            left,
+            right,
         }
     }
+}
+
+#[derive(
+    Debug,
+    PartialEq,
+    Eq,
+    Clone,
+    Serialize,
+    Deserialize,
+    JsonSchema,
+    derive_more::Display,
+    derive_more::From,
+    derive_more::TryInto,
+)]
+pub enum Identity<T> {
+    Polynomial(PolynomialIdentity<T>),
+    Lookup(LookupIdentity<T>),
+    Permutation(PermutationIdentity<T>),
+    Connect(ConnectIdentity<T>),
+}
+
+// TODO This is the only version of Identity left.
+impl<T> Identity<T> {
+    pub fn contains_next_ref(&self) -> bool {
+        self.children().any(|e| e.contains_next_ref())
+    }
 
     pub fn degree(&self) -> usize {
         self.children().map(|e| e.degree()).max().unwrap_or(0)
     }
+
+    pub fn id(&self) -> u64 {
+        match self {
+            Identity::Polynomial(i) => i.id,
+            Identity::Lookup(i) => i.id,
+            Identity::Permutation(i) => i.id,
+            Identity::Connect(i) => i.id,
+        }
+    }
+
+    pub fn source(&self) -> &SourceRef {
```

Comment:
There is a trait called SourceReference, I think we should implement that instead.

Reply:
Done